rust-query-crlite/src/main.rs

```diff
@@ -0,0 +1,794 @@
+            Ok(Some(sct_extension)) => sct_extension,
+            _ => return false,
+        };
+        let scts = match sct_extension.parsed_extension() {
+            ParsedExtension::SCT(scts) => scts,
+            _ => return false,
+        };
+        for sct in scts.iter() {
+            if let Some((min, max)) = self.0.get(sct.id.key_id.as_ref()) {
+                if *min <= sct.timestamp && sct.timestamp <= *max {
+                    debug!(
+                        "Logged at {} by enrolled log {}",
+                        sct.timestamp,
+                        base64::encode(sct.id.key_id)
+                    );
+                    return true;
+                }
+            }
+        }
+        false
+    }
+}
+
+struct Enrollment(HashSet<EnrollmentKey>);
+impl Enrollment {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, CRLiteDBError> {
+        let mut enrollment = HashSet::new();
+        let mut reader = BufReader::new(bytes);
+        match reader.read_u8() {
+            Ok(ENROLLMENT_SERIALIZATION_VERSION) => (),
+            _ => return Err(CRLiteDBError::from("unknown CRLite enrollment version")),
+        }
+        if (bytes.len() - 1) % ENROLLMENT_V1_ENTRY_BYTES != 0 {
+            return Err(CRLiteDBError::from("truncted CRLite enrollment file"));
+        }
+        let enrollment_count = (bytes.len() - 1) / ENROLLMENT_V1_ENTRY_BYTES;
+        for _ in 0..enrollment_count {
+            let mut enrollment_entry = [0u8; ENROLLMENT_V1_ENTRY_BYTES];
+            match reader.read_exact(&mut enrollment_entry) {
+                Ok(()) => (),
+                _ => return Err(CRLiteDBError::from("truncted CRLite enrollment file")),
+            };
+            let issuer_id = &enrollment_entry[..];
+            enrollment.insert(issuer_id.to_vec());
+        }
+        Ok(Enrollment(enrollment))
+    }
+
+    fn contains(&self, enrollment_key: &[u8]) -> bool {
+        self.0.contains(enrollment_key)
+    }
+}
+
+struct Intermediates(HashMap<IssuerDN, Vec<DERCert>>);
+impl Intermediates {
+    fn new() -> Self {
+        Intermediates(HashMap::new())
+    }
+
+    fn from_ccadb_csv(bytes: &[u8]) -> Result<Self, CRLiteDBError> {
+        // XXX: The CCADB report is a CSV file where the last entry in each logical line is a PEM
+        //      encoded cert. Unfortunately the newlines in the PEM encoding are not escaped, so
+        //      the logical line is split over several actual lines. Fortunately the pem crate is
+        //      happy to ignore content surrounding PEM data.
+        let list = pem::parse_many(bytes)
+            .map_err(|_| CRLiteDBError::from("error reading CCADB report"))?;
+
+        let mut intermediates = Intermediates::new();
+        for der in list {
+            if let Ok((_, cert)) = X509Certificate::from_der(&der.contents) {
+                let name = cert.tbs_certificate.subject.as_raw();
+                if let Some(existing) = intermediates.0.get_mut(name) {
```

Comment:
Could you do `intermediates.0.entry(name).or_insert_with(Vec::new)` and then append on to the result of that? Not a big deal, but it's similar to what you do above.